File: Linked List Cycle II/Linked_List_Cycle_II.cpp
// 双指针
// Runtime: 12 ms, faster than 29.54% of C++ online submissions for Linked List Cycle II.
// Memory Usage: 6.7 MB, less than 0.89% of C++ online submissions for Linked List Cycle II.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution 
{
public:
    ListNode *detectCycle(ListNode *head) 
    {
        if (head == 0)
            return 0;

        ListNode *fastPtr = head, *slowPtr = head;
        
        bool hasCycle = false;
        while (slowPtr->next != 0 && fastPtr->next != 0 && fastPtr->next->next !=0)
        {
            // 慢指针每次一共一格，快指针每次移动两格
            slowPtr = slowPtr->next;
            fastPtr = fastPtr->next->next;
            if (fastPtr == slowPtr)
            {
                hasCycle = true;
                break;
            }
        }

        if (hasCycle == false)
            return 0;

        // 此时链表中存在环
        fastPtr = head;
        while(slowPtr != fastPtr)
        {
            slowPtr = slowPtr->next;
            fastPtr = fastPtr->next;
        }
        return fastPtr;
    }
};

File: Linked List Cycle II/Linked_List_Cycle_II.py
# Runtime: 44 ms, faster than 99.63% of Python online submissions for Linked List Cycle II.
# Memory Usage: 16.2 MB, less than 94.55% of Python online submissions for Linked List Cycle II.

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head is None:
            return None
        fastPtr, slowPtr = head, head
        hasCycle = False
        while (slowPtr.next is not None and fastPtr.next is not None and fastPtr.next.next is not None):
            slowPtr = slowPtr.next
            fastPtr = fastPtr.next.next
            if slowPtr == fastPtr:
                hasCycle = True
                break
        if not hasCycle:
            return None
        fastPtr = head
        while fastPtr != slowPtr:
            slowPtr = slowPtr.next
            fastPtr = fastPtr.next
        return fastPtr